// @medusajs/workflow-engine-redis/dist/utils/workflow-orchestrator-storage.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _RedisDistributedTransactionStorage_instances, _RedisDistributedTransactionStorage_isWorkerMode, _RedisDistributedTransactionStorage_preventRaceConditionExecutionIfNecessary;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisDistributedTransactionStorage = void 0;
const orchestration_1 = require("@medusajs/framework/orchestration");
const utils_1 = require("@medusajs/framework/utils");
const bullmq_1 = require("bullmq");
var JobType;
(function (JobType) {
    JobType["SCHEDULE"] = "schedule";
    JobType["RETRY"] = "retry";
    JobType["STEP_TIMEOUT"] = "step_timeout";
    JobType["TRANSACTION_TIMEOUT"] = "transaction_timeout";
})(JobType || (JobType = {}));
class RedisDistributedTransactionStorage {
    constructor({ workflowExecutionService, redisConnection, redisWorkerConnection, redisQueueName, redisJobQueueName, logger, isWorkerMode, }) {
        _RedisDistributedTransactionStorage_instances.add(this);
        _RedisDistributedTransactionStorage_isWorkerMode.set(this, false);
        this.workflowExecutionService_ = workflowExecutionService;
        this.logger_ = logger;
        this.redisClient = redisConnection;
        this.redisWorkerConnection = redisWorkerConnection;
        this.queueName = redisQueueName;
        this.jobQueueName = redisJobQueueName;
        this.queue = new bullmq_1.Queue(redisQueueName, { connection: this.redisClient });
        this.jobQueue = isWorkerMode
            ? new bullmq_1.Queue(redisJobQueueName, {
                connection: this.redisClient,
            })
            : undefined;
        __classPrivateFieldSet(this, _RedisDistributedTransactionStorage_isWorkerMode, isWorkerMode, "f");
    }
    async onApplicationPrepareShutdown() {
        // Close worker gracefully, i.e. wait for the current jobs to finish
        await this.worker?.close();
        await this.jobWorker?.close();
    }
    async onApplicationShutdown() {
        await this.queue?.close();
        await this.jobQueue?.close();
    }
    async onApplicationStart() {
        const allowedJobs = [
            JobType.RETRY,
            JobType.STEP_TIMEOUT,
            JobType.TRANSACTION_TIMEOUT,
        ];
        const workerOptions = {
            connection: this.redisWorkerConnection /*, runRetryDelay: 100000 for tests */,
        };
        // TODO: Remove this once we have released to all clients (Added: v2.6+)
        // Remove all repeatable jobs from the old queue since now we have a queue dedicated to scheduled jobs
        await this.removeAllRepeatableJobs(this.queue);
        this.worker = new bullmq_1.Worker(this.queueName, async (job) => {
            this.logger_.debug(`executing job ${job.name} from queue ${this.queueName} with the following data: ${JSON.stringify(job.data)}`);
            if (allowedJobs.includes(job.name)) {
                await this.executeTransaction(job.data.workflowId, job.data.transactionId);
            }
            if (job.name === JobType.SCHEDULE) {
                // Remove repeatable job from the old queue since now we have a queue dedicated to scheduled jobs
                await this.remove(job.data.jobId);
            }
        }, workerOptions);
        if (__classPrivateFieldGet(this, _RedisDistributedTransactionStorage_isWorkerMode, "f")) {
            this.jobWorker = new bullmq_1.Worker(this.jobQueueName, async (job) => {
                this.logger_.debug(`executing scheduled job ${job.data.jobId} from queue ${this.jobQueueName} with the following options: ${JSON.stringify(job.data.schedulerOptions)}`);
                return await this.executeScheduledJob(job.data.jobId, job.data.schedulerOptions);
            }, workerOptions);
        }
    }
    setWorkflowOrchestratorService(workflowOrchestratorService) {
        this.workflowOrchestratorService_ = workflowOrchestratorService;
    }
    async saveToDb(data, retentionTime) {
        await this.workflowExecutionService_.upsert([
            {
                workflow_id: data.flow.modelId,
                transaction_id: data.flow.transactionId,
                execution: data.flow,
                context: {
                    data: data.context,
                    errors: data.errors,
                },
                state: data.flow.state,
                retention_time: retentionTime,
            },
        ]);
    }
    async deleteFromDb(data) {
        await this.workflowExecutionService_.delete([
            {
                workflow_id: data.flow.modelId,
                transaction_id: data.flow.transactionId,
            },
        ]);
    }
    async executeTransaction(workflowId, transactionId) {
        return await this.workflowOrchestratorService_.run(workflowId, {
            transactionId,
            logOnError: true,
            throwOnError: false,
        });
    }
    async executeScheduledJob(jobId, schedulerOptions) {
        try {
            // TODO: In the case of concurrency being forbidden, we want to generate a predictable transaction ID and rely on the idempotency
            // of the transaction to ensure that the transaction is only executed once.
            await this.workflowOrchestratorService_.run(jobId, {
                logOnError: true,
            });
        }
        catch (e) {
            if (e instanceof utils_1.MedusaError && e.type === utils_1.MedusaError.Types.NOT_FOUND) {
                this.logger_?.warn(`Tried to execute a scheduled workflow with ID ${jobId} that does not exist, removing it from the scheduler.`);
                await this.remove(jobId);
                return;
            }
            throw e;
        }
    }
    async get(key, options) {
        const data = await this.redisClient.get(key);
        if (data) {
            return JSON.parse(data);
        }
        const { idempotent, store, retentionTime } = options ?? {};
        if (!idempotent && !(store && (0, utils_1.isDefined)(retentionTime))) {
            return;
        }
        const [_, workflowId, transactionId] = key.split(":");
        const trx = await this.workflowExecutionService_
            .retrieve({
            workflow_id: workflowId,
            transaction_id: transactionId,
        }, {
            select: ["execution", "context"],
        })
            .catch(() => undefined);
        if (trx) {
            return {
                flow: trx.execution,
                context: trx.context.data,
                errors: trx.context.errors,
            };
        }
        return;
    }
    async list() {
        const keys = await this.redisClient.keys(orchestration_1.DistributedTransaction.keyPrefix + ":*");
        const transactions = [];
        for (const key of keys) {
            const data = await this.redisClient.get(key);
            if (data) {
                transactions.push(JSON.parse(data));
            }
        }
        return transactions;
    }
    async save(key, data, ttl, options) {
        /**
         * Store the retention time only if the transaction is done, failed or reverted.
         * From that moment, this tuple can be later on archived or deleted after the retention time.
         */
        const hasFinished = [
            utils_1.TransactionState.DONE,
            utils_1.TransactionState.FAILED,
            utils_1.TransactionState.REVERTED,
        ].includes(data.flow.state);
        const { retentionTime, idempotent } = options ?? {};
        await __classPrivateFieldGet(this, _RedisDistributedTransactionStorage_instances, "m", _RedisDistributedTransactionStorage_preventRaceConditionExecutionIfNecessary).call(this, {
            data,
            key,
            options,
        });
        if (hasFinished) {
            Object.assign(data, {
                retention_time: retentionTime,
            });
        }
        const stringifiedData = JSON.stringify(data);
        if (!hasFinished) {
            if (ttl) {
                await this.redisClient.set(key, stringifiedData, "EX", ttl);
            }
            else {
                await this.redisClient.set(key, stringifiedData);
            }
        }
        if (hasFinished && !retentionTime && !idempotent) {
            await this.deleteFromDb(data);
        }
        else {
            await this.saveToDb(data, retentionTime);
        }
        if (hasFinished) {
            await this.redisClient.unlink(key);
        }
    }
    async scheduleRetry(transaction, step, timestamp, interval) {
        await this.queue.add(JobType.RETRY, {
            workflowId: transaction.modelId,
            transactionId: transaction.transactionId,
            stepId: step.id,
        }, {
            delay: interval > 0 ? interval * 1000 : undefined,
            jobId: this.getJobId(JobType.RETRY, transaction, step),
            removeOnComplete: true,
        });
    }
    async clearRetry(transaction, step) {
        await this.removeJob(JobType.RETRY, transaction, step);
    }
    async scheduleTransactionTimeout(transaction, _, interval) {
        await this.queue.add(JobType.TRANSACTION_TIMEOUT, {
            workflowId: transaction.modelId,
            transactionId: transaction.transactionId,
        }, {
            delay: interval * 1000,
            jobId: this.getJobId(JobType.TRANSACTION_TIMEOUT, transaction),
            removeOnComplete: true,
        });
    }
    async clearTransactionTimeout(transaction) {
        await this.removeJob(JobType.TRANSACTION_TIMEOUT, transaction);
    }
    async scheduleStepTimeout(transaction, step, timestamp, interval) {
        await this.queue.add(JobType.STEP_TIMEOUT, {
            workflowId: transaction.modelId,
            transactionId: transaction.transactionId,
            stepId: step.id,
        }, {
            delay: interval * 1000,
            jobId: this.getJobId(JobType.STEP_TIMEOUT, transaction, step),
            removeOnComplete: true,
        });
    }
    async clearStepTimeout(transaction, step) {
        await this.removeJob(JobType.STEP_TIMEOUT, transaction, step);
    }
    getJobId(type, transaction, step) {
        const key = [type, transaction.modelId, transaction.transactionId];
        if (step) {
            key.push(step.id, step.attempts + "");
            if (step.isCompensating()) {
                key.push("compensate");
            }
        }
        return key.join(":");
    }
    async removeJob(type, transaction, step) {
        const jobId = this.getJobId(type, transaction, step);
        if (type === JobType.SCHEDULE) {
            const job = await this.jobQueue?.getJob(jobId);
            if (job) {
                await job.remove();
            }
        }
        else {
            const job = await this.queue.getJob(jobId);
            if (job && job.attemptsStarted === 0) {
                await job.remove();
            }
        }
    }
    /* Scheduler storage methods */
    async schedule(jobDefinition, schedulerOptions) {
        const jobId = typeof jobDefinition === "string" ? jobDefinition : jobDefinition.jobId;
        if ("cron" in schedulerOptions && "interval" in schedulerOptions) {
            throw new Error(`Unable to register a job with both scheduler options interval and cron.`);
        }
        const repeatOptions = {
            limit: schedulerOptions.numberOfExecutions,
            key: `${JobType.SCHEDULE}_${jobId}`,
        };
        if ("cron" in schedulerOptions) {
            repeatOptions.pattern = schedulerOptions.cron;
        }
        else {
            repeatOptions.every = schedulerOptions.interval;
        }
        // If it is the same key (eg. the same workflow name), the old one will get overridden.
        await this.jobQueue?.add(JobType.SCHEDULE, {
            jobId,
            schedulerOptions,
        }, {
            repeat: repeatOptions,
            removeOnComplete: {
                age: 86400,
                count: 1000,
            },
            removeOnFail: {
                age: 604800,
                count: 5000,
            },
        });
    }
    async remove(jobId) {
        await this.jobQueue?.removeRepeatableByKey(`${JobType.SCHEDULE}_${jobId}`);
    }
    async removeAll() {
        return await this.removeAllRepeatableJobs(this.jobQueue);
    }
    async removeAllRepeatableJobs(queue) {
        const repeatableJobs = (await queue.getRepeatableJobs()) ?? [];
        await (0, utils_1.promiseAll)(repeatableJobs.map((job) => queue.removeRepeatableByKey(job.key)));
    }
}
exports.RedisDistributedTransactionStorage = RedisDistributedTransactionStorage;
_RedisDistributedTransactionStorage_isWorkerMode = new WeakMap(), _RedisDistributedTransactionStorage_instances = new WeakSet(), _RedisDistributedTransactionStorage_preventRaceConditionExecutionIfNecessary = async function _RedisDistributedTransactionStorage_preventRaceConditionExecutionIfNecessary({ data, key, options, }) {
    let isInitialCheckpoint = false;
    if (data.flow.state === utils_1.TransactionState.NOT_STARTED) {
        isInitialCheckpoint = true;
    }
    /**
     * In case many execution can succeed simultaneously, we need to ensure that the latest
     * execution does continue if a previous execution is considered finished
     */
    const currentFlow = data.flow;
    const { flow: latestUpdatedFlow } = (await this.get(key, options)) ??
        { flow: {} };
    if (!isInitialCheckpoint && !(0, utils_1.isPresent)(latestUpdatedFlow)) {
        /**
         * the initial checkpoint expect no other checkpoint to have been stored.
         * In case it is not the initial one and another checkpoint is trying to
         * find if a concurrent execution has finished, we skip the execution.
         * The already finished execution would have deleted the checkpoint already.
         */
        throw new orchestration_1.SkipExecutionError("Already finished by another execution");
    }
    const currentFlowLastInvokingStepIndex = Object.values(currentFlow.steps).findIndex((step) => {
        return [
            utils_1.TransactionStepState.INVOKING,
            utils_1.TransactionStepState.NOT_STARTED,
        ].includes(step.invoke?.state);
    });
    const latestUpdatedFlowLastInvokingStepIndex = !latestUpdatedFlow.steps
        ? 1 // There is no other execution, so the current execution is the latest
        : Object.values(latestUpdatedFlow.steps ?? {}).findIndex((step) => {
            return [
                utils_1.TransactionStepState.INVOKING,
                utils_1.TransactionStepState.NOT_STARTED,
            ].includes(step.invoke?.state);
        });
    const currentFlowLastCompensatingStepIndex = Object.values(currentFlow.steps)
        .reverse()
        .findIndex((step) => {
        return [
            utils_1.TransactionStepState.COMPENSATING,
            utils_1.TransactionStepState.NOT_STARTED,
        ].includes(step.compensate?.state);
    });
    const latestUpdatedFlowLastCompensatingStepIndex = !latestUpdatedFlow.steps
        ? -1
        : Object.values(latestUpdatedFlow.steps ?? {})
            .reverse()
            .findIndex((step) => {
            return [
                utils_1.TransactionStepState.COMPENSATING,
                utils_1.TransactionStepState.NOT_STARTED,
            ].includes(step.compensate?.state);
        });
    const isLatestExecutionFinishedIndex = -1;
    const invokeShouldBeSkipped = (latestUpdatedFlowLastInvokingStepIndex ===
        isLatestExecutionFinishedIndex ||
        currentFlowLastInvokingStepIndex <
            latestUpdatedFlowLastInvokingStepIndex) &&
        currentFlowLastInvokingStepIndex !== isLatestExecutionFinishedIndex;
    const compensateShouldBeSkipped = currentFlowLastCompensatingStepIndex <
        latestUpdatedFlowLastCompensatingStepIndex &&
        currentFlowLastCompensatingStepIndex !== isLatestExecutionFinishedIndex &&
        latestUpdatedFlowLastCompensatingStepIndex !==
            isLatestExecutionFinishedIndex;
    if ((data.flow.state !== utils_1.TransactionState.COMPENSATING &&
        invokeShouldBeSkipped) ||
        (data.flow.state === utils_1.TransactionState.COMPENSATING &&
            compensateShouldBeSkipped) ||
        (latestUpdatedFlow.state === utils_1.TransactionState.COMPENSATING &&
            ![utils_1.TransactionState.REVERTED, utils_1.TransactionState.FAILED].includes(currentFlow.state) &&
            currentFlow.state !== latestUpdatedFlow.state) ||
        (latestUpdatedFlow.state === utils_1.TransactionState.REVERTED &&
            currentFlow.state !== utils_1.TransactionState.REVERTED) ||
        (latestUpdatedFlow.state === utils_1.TransactionState.FAILED &&
            currentFlow.state !== utils_1.TransactionState.FAILED)) {
        throw new orchestration_1.SkipExecutionError("Already finished by another execution");
    }
};
