// @medusajs/core-flows/dist/draft-order/workflows/begin-draft-order-edit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.beginDraftOrderEditWorkflow = exports.beginDraftOrderEditWorkflowId = void 0;
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const order_1 = require("../../order");
const steps_1 = require("../steps");
exports.beginDraftOrderEditWorkflowId = "begin-draft-order-edit";
exports.beginDraftOrderEditWorkflow = (0, workflows_sdk_1.createWorkflow)(exports.beginDraftOrderEditWorkflowId, function (input) {
    const order = (0, common_1.useRemoteQueryStep)({
        entry_point: "orders",
        fields: ["id", "status", "is_draft_order"],
        variables: { id: input.order_id },
        list: false,
        throw_if_key_not_found: true,
    }).config({ name: "order-query" });
    (0, steps_1.validateDraftOrderStep)({ order });
    const orderChangeInput = (0, workflows_sdk_1.transform)({ input }, ({ input }) => {
        return {
            change_type: "edit",
            order_id: input.order_id,
            created_by: input.created_by,
            description: input.description,
            internal_note: input.internal_note,
        };
    });
    return new workflows_sdk_1.WorkflowResponse((0, order_1.createOrderChangeStep)(orderChangeInput));
});
