// @medusajs/core-flows/dist/draft-order/workflows/convert-draft-order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertDraftOrderWorkflow = void 0;
const utils_1 = require("@medusajs/framework/utils");
const workflows_sdk_1 = require("@medusajs/framework/workflows-sdk");
const common_1 = require("../../common");
const validate_draft_order_1 = require("../steps/validate-draft-order");
const convertDraftOrderWorkflowId = "convert-draft-order";
/**
 * This step converts a draft order to a pending order.
 *
 * @example
 * ```typescript
 * const order = await convertDraftOrderStep({ id: "order_123" })
 * ```
 */
const convertDraftOrderStep = (0, workflows_sdk_1.createStep)("convert-draft-order", async function ({ id }, { container }) {
    const service = container.resolve(utils_1.Modules.ORDER);
    const response = await service.updateOrders([
        {
            id,
            status: utils_1.OrderStatus.PENDING,
            is_draft_order: false,
        },
    ]);
    const order = response[0];
    return new workflows_sdk_1.StepResponse(order, {
        id,
    });
}, async function (prevData, { container }) {
    if (!prevData) {
        return;
    }
    const service = container.resolve(utils_1.Modules.ORDER);
    await service.updateOrders([
        {
            id: prevData.id,
            status: utils_1.OrderStatus.DRAFT,
            is_draft_order: true,
        },
    ]);
});
/**
 * This workflow converts a draft order to a pending order.
 *
 * @example
 * ```typescript
 * const order = await convertDraftOrderWorkflow({ id: "order_123" })
 * ```
 */
exports.convertDraftOrderWorkflow = (0, workflows_sdk_1.createWorkflow)(convertDraftOrderWorkflowId, function (input) {
    const order = (0, common_1.useRemoteQueryStep)({
        entry_point: "orders",
        fields: ["id", "status", "is_draft_order"],
        variables: {
            id: input.id,
        },
        list: false,
        throw_if_key_not_found: true,
    }).config({ name: "order-query" });
    (0, validate_draft_order_1.validateDraftOrderStep)({ order });
    const updatedOrder = convertDraftOrderStep({ id: input.id });
    return new workflows_sdk_1.WorkflowResponse(updatedOrder);
});
